### PR TITLE
More Cargo improvements

### PR DESCRIPTION
**Pull request by sschuberth**
Please have a look at the individual commit messages for the details.

### CODE REVIEW

File: plugins/package-managers/cargo/src/main/kotlin/CargoMetadata.kt [hunk 1]
```diff
@@ -93,6 +93,31 @@ internal data class CargoMetadata(
...
+         * The dependencies of this package. This is an alternative to "dependencies" which contains additional
+         * information. In particular, this handles renamed dependencies.
+         */
+        val deps: List<Dep>
+    )
+
+    @Serializable
+    data class Dep(
+        /**
+         * The name of the dependency's library target. If this is a renamed dependency, this is the new name. */
```
Comment (mnonnenmacher) [inline plugins/package-managers/cargo/src/main/kotlin/CargoMetadata.kt:108]: Wrong linebreak after `/**`.

File: plugins/package-managers/cargo/src/main/kotlin/CargoMetadata.kt [hunk 2]
```diff
@@ -93,6 +93,31 @@ internal data class CargoMetadata(
         /** The Package ID of this node. */
         val id: String,
         /** The dependencies of this package, an array of Package IDs. */
-        val dependencies: List<String>
+        val dependencies: List<String>,
+
+        /**
+         * The dependencies of this package. This is an alternative to "dependencies" which contains additional
+         * information. In particular, this handles renamed dependencies.
```
Comment (mnonnenmacher) [inline plugins/package-managers/cargo/src/main/kotlin/CargoMetadata.kt:100]: Sounds to me like `dependencies` would contain additional information, but looking at the datatypes it seems that `deps` contains additional dependencies. Consider replacing "which" with "that".

File: plugins/package-managers/cargo/src/main/kotlin/CargoMetadata.kt [hunk 3]
```diff
@@ -93,6 +93,31 @@ internal data class CargoMetadata(
...
+        val name: String,
+        /** The Package ID of the dependency. */
+        val pkg: String,
+        /** Array of dependency kinds. Added in Cargo 1.40. */
+        val depKinds: List<DepKind>
+    )
+
+    @Serializable
+    data class DepKind(
+        /** The dependency kind. "dev", "build", or null for a normal dependency. */
```
Comment (mnonnenmacher) [inline plugins/package-managers/cargo/src/main/kotlin/CargoMetadata.kt:118]: Don't we use \`null\` (same below)?

File: plugins/package-managers/cargo/src/main/kotlin/Cargo.kt [hunk 4]
```diff
@@ -163,34 +138,43 @@ class Cargo(
             "Virtual workspaces are not supported."
         }
 
-        val projectNode = metadata.packages.single { it.id == projectId }
-        val groupedDependencies = projectNode.dependencies.groupBy { it.kind.orEmpty() }
+        val projectNode = metadata.resolve.nodes.single { it.id == projectId }
+        val depNodesByKind = mutableMapOf<String, MutableList<CargoMetadata.Node>>()
+        projectNode.deps.forEach { dep ->
+            dep.depKinds.forEach { depKind ->
+                val depNode = metadata.resolve.nodes.single { it.id == dep.pkg }
```
Comment (mnonnenmacher) [inline plugins/package-managers/cargo/src/main/kotlin/Cargo.kt:143]: Could this line be outside of this `forEach` loop?

File: plugins/package-managers/cargo/src/main/kotlin/Cargo.kt [hunk 5]
```diff
@@ -266,7 +266,8 @@ private fun parseSourceArtifact(pkg: CargoMetadata.Package, hashes: Map<String,
 
     if (source == "registry+https://github.com/rust-lang/crates.io-index") {
         val url = "https://crates.io/api/v1/crates/${pkg.name}/${pkg.version}/download"
-        val hash = Hash.create(hashes[pkg.id].orEmpty())
+        val key = "${pkg.name} ${pkg.version} (${pkg.source})"
+        val hash = Hash.create(hashes[key].orEmpty())
```
Comment (mnonnenmacher) [inline plugins/package-managers/cargo/src/main/kotlin/Cargo.kt:260]: Commit message: Why not?

File: plugins/package-managers/cargo/src/main/kotlin/Cargo.kt
Comment (mnonnenmacher) [inline plugins/package-managers/cargo/src/main/kotlin/Cargo.kt:1]: Commit message: Typo in "property"

File: plugins/package-managers/cargo/src/main/kotlin/Cargo.kt [hunk 6]
```diff
@@ -199,15 +189,15 @@ class Cargo(
             scopeDependencies = scopes
         )
 
-        val nonProjectPackages = packages
-            .filterNot { isProjectDependency(it.key) }
-            .mapTo(mutableSetOf()) { it.value }
+        val nonProjectPackages = packageById.values.mapNotNullTo(mutableSetOf()) { cargoPkg ->
+            cargoPkg.takeUnless { it.isProject() }?.let { parsePackage(cargoPkg, hashes) }
```
Comment (mnonnenmacher) [inline plugins/package-managers/cargo/src/main/kotlin/Cargo.kt]: Nit: Could be `parsePackage(it, hashes)`, fits better to the usage of `let`.

File: plugins/package-managers/cargo/src/main/kotlin/Cargo.kt [hunk 7]
```diff
@@ -199,15 +189,15 @@ class Cargo(
...
+        }
 
         return listOf(ProjectAnalyzerResult(project, nonProjectPackages))
     }
 }
 
-private val PATH_DEPENDENCY_REGEX = Regex("""^.*\(path\+file://(.*)\)$""")
+private fun CargoMetadata.Package.isProject(): Boolean = source == null
```
Comment (mnonnenmacher) [inline plugins/package-managers/cargo/src/main/kotlin/Cargo.kt:200]: Unnecessary return type declaration.

File: plugins/package-managers/cargo/src/main/kotlin/Cargo.kt [hunk 8]
```diff
@@ -190,7 +190,7 @@ class Cargo(
         )
 
         val nonProjectPackages = packageById.values.mapNotNullTo(mutableSetOf()) { cargoPkg ->
-            cargoPkg.takeUnless { it.isProject() }?.let { parsePackage(cargoPkg, hashes) }
+            cargoPkg.takeUnless { it.isProject() }?.let { cargoPkg.toPackage(hashes) }
```
Comment (mnonnenmacher) [inline plugins/package-managers/cargo/src/main/kotlin/Cargo.kt:193]: `let` is now unnecessary.

**Review by mnonnenmacher — CHANGES_REQUESTED** [review]

File: plugins/package-managers/cargo/src/main/kotlin/CargoMetadata.kt (hunk 2 above)
Comment (sschuberth) [reply plugins/package-managers/cargo/src/main/kotlin/CargoMetadata.kt:100]: I've taken all of these comments as-is from [here](https://doc.rust-lang.org/cargo/commands/cargo-metadata.html#json-format), so I'd prefer not to change them, or keep changes to a minimum.

File: plugins/package-managers/cargo/src/main/kotlin/CargoMetadata.kt (hunk 3 above)
Comment (sschuberth) [reply plugins/package-managers/cargo/src/main/kotlin/CargoMetadata.kt:118]: See above; I'm undecided whether this is a formatting change that I still should do or not.

File: plugins/package-managers/cargo/src/main/kotlin/Cargo.kt (hunk 5 above)
Comment (sschuberth) [reply plugins/package-managers/cargo/src/main/kotlin/Cargo.kt:260]: I've explained it in the commit message now.

**Review by fviernau — CHANGES_REQUESTED** [review]

File: plugins/package-managers/cargo/src/main/kotlin/CargoMetadata.kt [hunk 9]
```diff
@@ -93,6 +93,30 @@ internal data class CargoMetadata(
...
+        val kind: String? = null,
+        /** The target platform for the dependency. null if not a target dependency. */
```
[the remark this answers is not in the document]
Comment (sschuberth) [reply plugins/package-managers/cargo/src/main/kotlin/CargoMetadata.kt:119]: Please see [above](https://github.com/oss-review-toolkit/ort/pull/8497#discussion_r1556944075), if possible I prefer to stick to the exact wording of the official docs.

File: plugins/package-managers/cargo/src/main/kotlin/Cargo.kt (hunk 6 above)
Comment (mnonnenmacher) [reply plugins/package-managers/cargo/src/main/kotlin/Cargo.kt]: It's obsolete with the following change anyway.

**Pull request merged**